### PR TITLE
Add timeout for maximum time spent running rule on file

### PR DESCRIPTION
**Pull request by brendongo**

### CODE REVIEW

**Comment by brendongo** [comment]
Relevant: https://github.com/returntocorp/semgrep/issues/1361

File: semgrep/semgrep/core_exception.py [hunk 1]
```diff
@@ -56,27 +59,38 @@ def from_json(  # type: ignore
         start_pos = Position(start["line"], start["col"])
         end_pos = Position(end["line"], end["col"])
 
+        # Semgrep-Core caches TimeoutErrors as FatalErrors
+        # Hack to treat these as TimeoutErrors
+        check_id = json_obj["check_id"]
+        if check_id == "FatalError" and "Timeout" in json_obj["extra"]["message"]:
```
Comment (nbrahms) [inline semgrep/semgrep/core_exception.py:65]: lol

File: semgrep/semgrep/core_exception.py [hunk 2]
```diff
@@ -56,27 +59,38 @@ def from_json(  # type: ignore
...
+            check_id = "Timeout"
+
         return cls(
-            json_obj["check_id"],
+            check_id,
```
Comment (nbrahms) [inline semgrep/semgrep/core_exception.py:69]: how do you feel about `json_obj.get("check_id", rule_id)`?

File: semgrep/semgrep/core_runner.py [hunk 3]
```diff
@@ -468,18 +473,8 @@ def dedup_output(outputs: List[RuleMatch]) -> List[RuleMatch]:
...
-            str(error._path),
-            error._start.line,
-            error._start.col,
-            error._end.line,
-            error._end.col,
-        )
-
-    return list({uniq_error_id(r): r for r in errors}.values())
+def dedup_errors(errors: List[SemgrepError]) -> List[SemgrepError]:
+    return list(set(errors))
```
Comment (nbrahms) [inline semgrep/semgrep/core_runner.py:477]: I assume `hash` is set up on `SemgrepError` to do the right thing here?

**Review by nbrahms — APPROVED** [review]

File: semgrep/semgrep/core_runner.py (hunk 3 above)
Comment (brendongo) [reply semgrep/semgrep/core_runner.py:477]: Yep. Attrs handles it correctly and there are tests verifying it

File: semgrep/semgrep/core_exception.py (hunk 2 above)
Comment (brendongo) [reply semgrep/semgrep/core_exception.py:69]: This is needed for the above hack 😢

**Comment by aryx** [comment]
Nice!

**Comment by brendongo** [comment]
Also related: https://github.com/returntocorp/semgrep/issues/1368

**Pull request merged**